internal/plugin/connectors/http/generic/oauth/v1/protocol.go

```diff
@@ -0,0 +1,222 @@
+		SignatureMethod: oauth_signature_method_type,
+		Token:           params[oauth_token],
+		TokenSecret:     params[oauth_token_secret],
+		TimeStamp:       strconv.Itoa(int(time.Now().Unix())),
+		Version:         oauth_version_number,
+	}
+
+	return oauth
+}
+
+// encodeURI percent encodes a string and changes encoding of
+// <SPACE> from "+" to "%20"
+// https://tools.ietf.org/html/rfc5849#section-3.6
+func encodeURI(stringToEncode string) string {
+	return strings.ReplaceAll(url.QueryEscape(stringToEncode), "+", "%20")
+}
+
+// collectParameters creates a map of body and query parameters by
+// percent encoding all keys and values
+// and sorting values post-encoding
+func collectParameters(oauth1 OAuth1, r *gohttp.Request) map[string][]string {
+	// Create two copies of the request body
+	// Since r.Body is io.ReadCloser the body is empty after reading.
+	// Need to replace r.Body after r.ParseForm() dumps the body contents
+	buf, _ := ioutil.ReadAll(r.Body)
+	originalBody := ioutil.NopCloser(bytes.NewBuffer(buf))
+	copyBody := ioutil.NopCloser(bytes.NewBuffer(buf))
+	r.Body = originalBody
+
+	paramMap := make(map[string][]string, 0)
+	// get query and body params and URL
+	// encode keys and values
+	// sort values after encoding
+	r.ParseForm()
+	for paramKey, paramValues := range r.Form {
+		encodedKey := encodeURI(paramKey)
+		paramMap[encodedKey] = paramValues
+
+		// URL encode all values in all keys
+		// Keys with multiple values require the values to be sorted post-percent encoding
+		for valueIndex, value := range paramMap[encodedKey] {
+			encodedValue := encodeURI(value)
+			paramMap[encodedKey][valueIndex] = encodedValue
+		}
+		sort.Strings(paramMap[encodedKey])
+	}
+	r.Body = copyBody
+
+	// these keys are already URL safe and don't need sorted(1 value)
+	paramMap[oauth_consumer_key] = []string{encodeURI(oauth1.ConsumerKey)}
+	paramMap[oauth_nonce] = []string{encodeURI(oauth1.Nonce)}
+	paramMap[oauth_signature_method] = []string{encodeURI(oauth1.SignatureMethod)}
+	paramMap[oauth_timestamp] = []string{encodeURI(oauth1.TimeStamp)}
+	paramMap[oauth_token] = []string{encodeURI(oauth1.Token)}
+	paramMap[oauth_version] = []string{encodeURI(oauth1.Version)}
+
+	return paramMap
+}
+
+// generateParameterString creates a Parameter String from sorting an encoded map of parameters and values
+// https://tools.ietf.org/html/rfc5849#section-3.5
+func generateParameterString(paramMap map[string][]string) string {
+	var parameterString bytes.Buffer
+
+	// make map of keys and sort to append lexicographically later
+	keys := make([]string, 0, len(paramMap))
+	for k := range paramMap {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	// create string
+	for keyIndex, key := range keys {
+		// Write keys and value pairs to `parameterString`
+		// ex: key0=value&key1=value&key2=value
+		for valueIndex, value := range paramMap[key] {
+			// Do not add '&' on first value
+			if keyIndex != 0 || valueIndex > 0 {
```

Comment:
These two conditions seem the same but are written differently

Reply:
If the first key has multiple values, we want to append the ampersand for the second value.

Reply:
And we never want to add the ampersand before the first key/we want to append the ampersand for multiple keys

So this takes care of this case:
```
key[]=value1&key[]=value2
```
where `valueIndex` is 1 for `value2`

This also takes care of this case:
```
key1=value1&key2=value2
```
we only want to add the ampersand after `key1`

Reply:
I'm not sure I understand - the comment was just about making the condition look the same:
```go
if keyIndex > 0 || valueIndex > 0 {
```